### PR TITLE
[Feature] Integrated data source provider into the form builder

### PR DESCRIPTION
**Pull request by YaroslavOvdii**
@sofiiakvasnevska 

## Issue
https://weboo.atlassian.net/browse/OD-10

## Description
Integrated data source provider into the form builder

## Screenshots/screencasts
https://share.getcloudapp.com/7KuLKQLx

## Backward compatibility
This change is fully backward compatible.

## Reviewers
@upplabs-alex-levchenko

### CODE REVIEW

File: js/libs/builder.js [hunk 1]
```diff
@@ -686,6 +629,47 @@ var app = new Vue({
 
       return string;
     },
+    initDataSourceProvider: function() {
+      var $vm = this;
+      var dataSourceData = {
+        dataSourceTitle: 'Form builder data source',
```
Comment (tonytlwu) [inline js/libs/builder.js:635]:
```suggestion
        dataSourceTitle: 'Form data source',
```

File: js/libs/builder.js [hunk 2]
```diff
@@ -686,6 +629,47 @@ var app = new Vue({
...
+        dataSourceId: $vm.settings.dataSourceId,
+        appId: Fliplet.Env.get('appId'),
+        default: {
+          name: 'Form builder data for ' + Fliplet.Env.get('appName'),
+          definition: {
```
Comment (tonytlwu) [inline js/libs/builder.js:640]: This doesn't need a `guid`.

File: js/libs/builder.js [hunk 3]
```diff
@@ -686,6 +629,47 @@ var app = new Vue({
...
       return string;
     },
+    initDataSourceProvider: function() {
+      var $vm = this;
+      var dataSourceData = {
+        dataSourceTitle: 'Form builder data source',
+        dataSourceId: $vm.settings.dataSourceId,
+        appId: Fliplet.Env.get('appId'),
+        default: {
+          name: 'Form builder data for ' + Fliplet.Env.get('appName'),
```
Comment (tonytlwu) [inline js/libs/builder.js:639]:
```suggestion
          name: 'Form data for ' + formName,
```

File: js/libs/builder.js [hunk 4]
```diff
@@ -686,6 +629,47 @@ var app = new Vue({
...
+          definition: {
+            'guid': 'Column 1'
+          },
+          entries: [],
+          columns: []
+        },
+        accessRules: [
```
Comment (tonytlwu) [inline js/libs/builder.js:646]: Incorrect access rules. See https://docs.google.com/presentation/d/1vvEvfGm_XGjKE3vp1duMy8DuqrgQGD3Dvlf4M-xNQQU/edit?ts=5f341a02#slide=id.g8d5e20b4f1_0_0

@upplabs-alex-levchenko @sofiiakvasnevska Please follow the specs provided. I shouldn't need to open the document to check this myself.

**Review by tonytlwu — CHANGES_REQUESTED** [review]
@squallstar Could you review this too as you were involved in the initial build of this component?

File: js/libs/builder.js (hunk 4 above)
Comment (YaroslavOvdii) [reply js/libs/builder.js:646]: @tonytlwu it said in the spec that form builder rules are: `Write, Update - only if the update is enabled` and that is exactly how rules are set. Is there something more that should be added to these rules?
image

File: js/libs/builder.js (hunk 2 above)
Comment (YaroslavOvdii) [reply js/libs/builder.js:640]: I'll remove it then.

File: js/libs/builder.js (hunk 4 above)
Comment (tonytlwu) [reply js/libs/builder.js:643]: @YaroslavOvdii Yes, that is correct. Sorry this was my bad, I saw `"allow": "all"` and it looked like we're allowing them to do everything, but I was mistaken. This is fine.

**Comment by tonytlwu** [comment]
@YaroslavOvdii Looks good to me now. I'd like @squallstar to review and give it the OK too.

File: js/libs/builder.js [hunk 5]
```diff
@@ -144,7 +144,8 @@ var app = new Vue({
       showDataSourceSettings: !!formSettings.dataSourceId,
       organizationName: '',
       isPreviewing: formSettings.previewingTemplate !== '',
-      editor: undefined
+      editor: undefined,
+      accessRulesTypes: ['insert']
```
Comment (squallstar) [inline js/libs/builder.js:148]: what's this for and why doesn't it have "update" too?

File: js/libs/builder.js [hunk 6]
```diff
@@ -686,6 +629,44 @@ var app = new Vue({
...
       return string;
     },
+    initDataSourceProvider: function() {
+      var $vm = this;
+      var dataSourceData = {
+        dataSourceTitle: 'Form data source',
+        dataSourceId: $vm.settings.dataSourceId,
+        appId: Fliplet.Env.get('appId'),
+        default: {
+          name: 'Form data for ' + $vm.settings.displayName,
```
Comment (squallstar) [inline js/libs/builder.js:639]: @tonytlwu has this name been agreed?

File: js/libs/builder.js [hunk 7]
```diff
@@ -832,6 +809,14 @@ var app = new Vue({
     'settings.dataStore': function(value) {
       this.showExtraAdd = value.indexOf('dataSource') > -1;
       this.showExtraEdit = value.indexOf('editDataSource') > -1;
```
Comment (squallstar) [inline js/libs/builder.js:811]: what does showExtraEdit mean? I also don't see it defined in the data object

**Review by squallstar — CHANGES_REQUESTED** [review]
Reviewed

File: js/libs/builder.js (hunk 6 above)
Comment (tonytlwu) [reply js/libs/builder.js:639]: Yes. I brought it up in a previous comment. It's been updated based on my suggestions.

File: js/libs/builder.js (hunk 7 above)
Comment (YaroslavOvdii) [reply js/libs/builder.js:811]: @squallstar If I understand it right it is showing an extra setting when we click on ` Update existing entries in a data source` checkbox.
image
Same thing for `showExtraAdd `.

File: js/libs/builder.js (hunk 5 above)
Comment (YaroslavOvdii) [reply js/libs/builder.js:148]: @squallstar I've added the `accessRulesTypes` variable so we could change its value depending on what setting is set.
In this [file](https://docs.google.com/presentation/d/1vvEvfGm_XGjKE3vp1duMy8DuqrgQGD3Dvlf4M-xNQQU/edit#slide=id.g8d5e20b4f1_0_0) said that we should set the data source rule `update` only if this `Update existing entries in a data source` checkbox is selected. That is why when we click on it we change our variable in what state it should be. You may see it at [this](https://github.com/Fliplet/fliplet-widget-form-builder/pull/330/files#diff-892056c1c88dfac81f33ecf67625a5ecR812) line.

**Review by squallstar — APPROVED** [review]

**Comment by squallstar** [comment]
Approved.

**Review by tonytlwu — APPROVED** [review]
@sofiiakvasnevska This is ready for testing

Risk: Critical
Affected: Form components configured in Studio
Verify: Data configurations are correctly saved and loaded

**Pull request merged**